src/raft/raft.rs

```diff
@@ -1764,26 +1822,50 @@ impl<T: Storage> Raft<T> {
     // promotable indicates whether state machine can be promoted to leader,
     // which is true when its own id is in progress list.
     pub fn promotable(&self) -> bool {
-        self.get_prs().contains_key(&self.id)
+        self.get_prs().voters().contains_key(&self.id)
     }
 
-    pub fn add_node(&mut self, id: u64) {
+    fn add_voter_or_learner(&mut self, id: u64, is_learner: bool) {
```

Comment:
Can you keep the logic in this function be same as etcd? The rewrite have introduced several problems already.